// candycane/dist/foundation/application.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _loader = require('./loader');

var _loader2 = _interopRequireDefault(_loader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let Application = class Application {
  constructor() {
    let options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    Object.assign(this, options);

    this.loader = options.loader || new _loader2.default(options.projectDir);
  }

  singleton(name, instance) {
    this.loader.register(name, instance);
  }

  make(name) {
    return this.loader.make(name);
  }

  boot() {
    this.config.providers.forEach(provider => {
      this.runProvider(provider);
    });
  }

  runProvider(curr) {
    const module = this.loader.requireNpm(curr);
    const klass = module.default || module;

    const provider = new klass(this);

    provider.register();
  }

  pathsForNamespace(namespace) {
    return this.loader.pathsForNamespace(namespace);
  }
};
exports.default = Application;
